build.js

```diff
@@ -27,24 +30,59 @@ await build({
     {
       name: 'alias',
       setup({ onResolve, resolve }) {
-        onResolve({ filter: /^fs$/ }, () => ({
-          path: fileURLToPath(new URL('src/stubs/fs.cjs', import.meta.url)),
-          sideEffects: false,
-        }));
-        onResolve({ filter: /^util$/ }, () => ({
-          path: fileURLToPath(new URL('src/stubs/util.cjs', import.meta.url)),
-          sideEffects: false,
-        }));
-        onResolve({ filter: /^vscode-emmet-helper-bundled$/ }, () => ({
-          path: fileURLToPath(new URL('src/stubs/noop.cjs', import.meta.url)),
+        // These packages are imported, but can be stubbed.
+        onResolve(
+          {
+            filter:
+              /^(detect-indent|chalk|fs|path|util|util-deprecate|vscode-emmet-helper-bundled)$/,
+          },
+          ({ path }) => ({
+            path: fileURLToPath(new URL(`src/stubs/${path}.ts`, import.meta.url)),
+            sideEffects: false,
+          }),
+        );
+
+        // The tailwindcss main export exports CJS, but we can get better tree shaking if we import
+        // from the ESM src directoy instead.
+        onResolve({ filter: /^tailwindcss$/ }, ({ path, ...options }) =>
+          resolve('tailwindcss/src', options),
+        );
+        onResolve({ filter: /^tailwindcss($|\/lib)/ }, ({ path, ...options }) =>
+          resolve(path.replace('lib', 'src'), options),
+        );
+
+        // The tailwindcss-language-service main export exports CJS by default, but we get better
+        // tree shaking if we import the ESM variant.
+        onResolve({ filter: /^tailwindcss-language-service$/ }, ({ path, ...options }) =>
+          resolve('tailwindcss-language-service/dist/tailwindcss-language-service.esm.js', options),
+        );
+
+        // This file pulls in a number of dependencies, but we don’t really need it anyway.
+        onResolve({ filter: /^\.+\/(util\/)?log$/ }, ({ path, ...options }) =>
+          options.importer.includes(`${sep}tailwindcss${sep}`)
+            ? {
+                path: fileURLToPath(new URL('src/stubs/tailwindcss/utils/log.ts', import.meta.url)),
+                sideEffects: false,
+              }
+            : resolve(path, options),
```

Comment:
recursion detector says beeeep :D

(onResolve will be triggered again via resolve https://github.com/evanw/esbuild/issues/2032)

Reply:
Hmm you’re right! So I guess we should throw an error in this case?

Reply:
thats the best solution that came to my mind too: https://github.com/remcohaszing/monaco-tailwindcss/pull/21/files#diff-7d111b05fe05c50beb729d1e99a36010fbda8c89b57a1a26ac3e7b0778a0ed53R35

it would be an evil idea now to use require.resolve as this functionality doesn't equal esbuilds resolve -> this would make it terrible to debug ^^ better an exception and then figure stuff out!